File: ports/php/tests/ModelTest.php
<?php

declare(strict_types=1);

namespace Lombok\TableSheet\Tests;

use Lombok\TableSheet\Core\Sheet;
use Lombok\TableSheet\Core\Workbook;
use PHPUnit\Framework\TestCase;

final class ModelTest extends TestCase
{
    public function testSetCellGetCellRoundTrip(): void
    {
        $sheet = new Sheet('S1');
        $sheet->setValue(0, 0, 'Name');
        $sheet->setValue(0, 1, 'Age');
        $sheet->setValue(1, 0, 'Alice');
        $sheet->setValue(1, 1, 30);

        $this->assertSame('Name', $sheet->getCell(0, 0)->value);
        $this->assertSame(30, $sheet->getCell(1, 1)->value);
        $this->assertSame('number', $sheet->getCell(1, 1)->type);
        $this->assertSame(2, $sheet->rowCount);
        $this->assertSame(2, $sheet->colCount);
    }

    public function testGetCellOnUnsetCellReturnsEmpty(): void
    {
        $sheet = new Sheet('S1');
        $cell = $sheet->getCell(5, 5);
        $this->assertNull($cell->value);
        $this->assertSame('empty', $cell->type);
    }

    public function testSetCellRejectsNegativeCoordinates(): void
    {
        $this->expectException(\RangeException::class);
        (new Sheet('S1'))->setCell(-1, 0, new \Lombok\TableSheet\Core\Cell(1, 'number'));
    }

    public function testWorkbookFromRowsBuildsCorrectDimensions(): void
    {
        $wb = Workbook::fromRows([
            ['a', 'b', 'c'],
            [1, 2, 3],
        ]);
        $sheet = $wb->sheets[0];
        $this->assertSame(2, $sheet->rowCount);
        $this->assertSame(3, $sheet->colCount);
        $this->assertSame([['a', 'b', 'c'], [1, 2, 3]], $sheet->toRows());
    }

    public function testSheetCloneIsDeepAndIndependent(): void
    {
        $sheet = new Sheet('S1');
        $sheet->setValue(0, 0, 'x');
        $clone = $sheet->clone();
        $clone->setValue(0, 0, 'y');
        $this->assertSame('x', $sheet->getCell(0, 0)->value);
        $this->assertSame('y', $clone->getCell(0, 0)->value);
    }
}

File: ports/go/lombok/model_test.go
package lombok

import (
	"reflect"
	"testing"
)

func TestSheetSetGetCellRoundTrip(t *testing.T) {
	s := NewSheet("S1")
	_ = s.SetValue(0, 0, "Name")
	_ = s.SetValue(0, 1, "Age")
	_ = s.SetValue(1, 0, "Alice")
	_ = s.SetValue(1, 1, 30.0)

	if v := s.GetCell(0, 0).Value; v != "Name" {
		t.Errorf("expected Name, got %v", v)
	}
	if v := s.GetCell(1, 1).Value; v != 30.0 {
		t.Errorf("expected 30, got %v", v)
	}
	if s.GetCell(1, 1).Type != CellNumber {
		t.Errorf("expected CellNumber type")
	}
	if s.RowCount != 2 || s.ColCount != 2 {
		t.Errorf("expected 2x2 dims, got %dx%d", s.RowCount, s.ColCount)
	}
}

func TestGetCellOnUnsetCellReturnsEmpty(t *testing.T) {
	s := NewSheet("S1")
	c := s.GetCell(5, 5)
	if c.Value != nil || c.Type != CellEmpty {
		t.Errorf("expected empty cell, got %+v", c)
	}
}

func TestSetCellRejectsNegativeCoordinates(t *testing.T) {
	s := NewSheet("S1")
	err := s.SetCell(-1, 0, Cell{Value: 1.0, Type: CellNumber})
	if err == nil {
		t.Fatal("expected an error for negative coordinates")
	}
}

func TestWorkbookFromRowsBuildsCorrectDimensions(t *testing.T) {
	w := WorkbookFromRows([][]interface{}{
		{"a", "b", "c"},
		{1.0, 2.0, 3.0},
	}, "", "")
	sheet := w.Sheets[0]
	if sheet.RowCount != 2 || sheet.ColCount != 3 {
		t.Errorf("expected 2x3, got %dx%d", sheet.RowCount, sheet.ColCount)
	}
	got := sheet.ToRows()
	want := [][]interface{}{{"a", "b", "c"}, {1.0, 2.0, 3.0}}
	if !reflect.DeepEqual(got, want) {
		t.Errorf("got %v, want %v", got, want)
	}
}

func TestSheetCloneIsDeepAndIndependent(t *testing.T) {
	s := NewSheet("S1")
	_ = s.SetValue(0, 0, "x")
	clone := s.Clone()
	_ = clone.SetValue(0, 0, "y")
	if s.GetCell(0, 0).Value != "x" {
		t.Errorf("original should be unaffected, got %v", s.GetCell(0, 0).Value)
	}
	if clone.GetCell(0, 0).Value != "y" {
		t.Errorf("clone should have new value, got %v", clone.GetCell(0, 0).Value)
	}
}

func TestResizeShrinksDimensionsAndToRowsRespectsIt(t *testing.T) {
	// Regression test mirroring the real bug found in the TS core via fuzz
	// testing (see SECURITY.md): shrinking dimensions must not leak stale
	// cells back into ToRows() as jagged rows.
	s := NewSheet("S1")
	_ = s.SetValue(0, 0, "a")
	_ = s.SetValue(1, 1, "b")
	_ = s.SetValue(2, 2, "c")
	if s.RowCount != 3 || s.ColCount != 3 {
		t.Fatalf("expected 3x3 before resize, got %dx%d", s.RowCount, s.ColCount)
	}
	s.Resize(1, 1)
	rows := s.ToRows()
	if len(rows) != 1 || len(rows[0]) != 1 {
		t.Fatalf("expected a single 1x1 row after resize, got %v", rows)
	}
	if rows[0][0] != "a" {
		t.Errorf("expected surviving cell 'a', got %v", rows[0][0])
	}
}
